### PR TITLE
[StaticWebAssets] Detects pre-compressed assets

### PR DESCRIPTION
**Pull request by javiercn**
# Detects pre-compressed assets

Detects assets that have been pre-compressed by an external tool to avoid conflicts when we compress our own assets, and they end up on the same path as the existing pre-compressed assets.

## Description

We have received feedback that some customers are facing issues upgrading as they are consuming packages that contain pre-compressed assets or are using third-party tools (like webpack) that produce pre-compressed versions of those assets.

Our recommendation in these cases is to let the framework handle the compression, as it further optimizes the delivery of those assets. However, this requires changes on their project to disable the compression on the third-party tools / exclude the pre compressed assets or to disable the compression that is done by the framework (a one-line MSBuild change).

We, however, recognize that this cause friction during the upgrade process and want to improve the situation by following the most common convention, which is, detecting compressed versions of an asset that live side-by-side with it. (Same path with an extra .gz or .br for the compressed versions)

This change detects such scenarios and starts treating the assets in the same way as if they were defined by the framework.

Fixes https://github.com/dotnet/aspnetcore/issues/57518

## Customer Impact

Customers with pre compressed files in their web content are forced to either remove those assets from the build or disable compression when they upgrade, as otherwise the build breaks.

## Regression?

- [ ] Yes
- [X] No

It's not a regression because this is a new feature, but it impacts the upgrade flow in this particular scenario.

## Risk

- [ ] High
- [ ] Medium
- [X] Low

There is a flag that can be used to turn off compression on the framework, which will prevent any compression related code from running.

The new added logic should no-op in most common cases and only ever execute when we detect a pre-compressed asset that was not generated by the framework.

## Verification

- [ ] Manual (required)
- [X] Automated

## Packaging changes reviewed?

- [ ] Yes
- [ ] No
- [X] N/A

----

## When servicing release/2.1

- [ ] Make necessary changes in eng/PatchConfig.props

### CODE REVIEW

File: src/StaticWebAssetsSdk/Targets/Microsoft.NET.Sdk.StaticWebAssets.Compression.targets [hunk 1]
```diff
@@ -276,6 +277,27 @@ Copyright (c) .NET Foundation. All rights reserved.
...
+      Filters=""
+    >
+      <Output TaskParameter="FilteredEndpoints" ItemName="_PrecompressedEndpointsToRemove" />
+    </FilterStaticWebAssetEndpoints>
+
+    <ItemGroup Condition="'@(_PrecompressedStaticWebAssets)' != ''">
+      <StaticWebAssetEndpoint Remove="@(_PrecompressedEndpointsToRemove)" />
+      <StaticWebAsset Remove="@(_PrecompressedStaticWebAssets)" />
+      <StaticWebAsset Include="@(_PrecompressedStaticWebAssets)" />
+    </ItemGroup>
```
Comment (javiercn) [inline src/StaticWebAssetsSdk/Targets/Microsoft.NET.Sdk.StaticWebAssets.Compression.targets:300]: * Discover existing pre-compressed assets.
* Remove their original definitions.
* Provide an updated definition.
* Remove the associated endpoint definitions and re-generate new endpoints.

**Review by MackinnonBuck — APPROVED** [review]

File: src/StaticWebAssetsSdk/Tasks/Compression/DiscoverPrecompressedAssets.cs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+                "Skipping task '{0}' because no candidate assets for compression were specified.",
+                nameof(ResolveCompressedAssets));
+            return true;
+        }
+
+        var candidates = CandidateAssets.Select(StaticWebAsset.FromTaskItem).ToArray();
+        var assetsToUpdate = new List<ITaskItem>();
+
+        // This is a loop inside of a loop (n^2), however, we hope the first condition is a good filter to
+        // avoid the inner loop in most cases.
```
[the remark this answers is not in the document]
Comment (javiercn) [reply src/StaticWebAssetsSdk/Tasks/Compression/DiscoverPrecompressedAssets.cs:35]: I've switched this to check against a dictionary.

File: src/StaticWebAssetsSdk/Tasks/Compression/DiscoverPrecompressedAssets.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+using System.Security.Cryptography;
+using Microsoft.Build.Framework;
+using Microsoft.Extensions.FileSystemGlobbing;
+
+namespace Microsoft.AspNetCore.StaticWebAssets.Tasks;
+
+public class DiscoverPrecompressedAssets : Task
+{
+    private static readonly char[] PatternSeparator = [';'];
```
[the remark this answers is not in the document]
Comment (javiercn) [reply src/StaticWebAssetsSdk/Tasks/Compression/DiscoverPrecompressedAssets.cs]: Removed

**Review by MattParkerDev — APPROVED** [review]

**Comment by mkArtakMSFT** [comment]
I've got verbal approval about this one (@marcpopMSFT , @SteveMCarroll )

**Pull request merged**